File: Lumo/Services/Announcement.swift
import Foundation
import UIKit

/// Ana menüde gösterilen duyuru kartı — "yeni sürüm çıktı", "bu hafta çift
/// yıldız" gibi.
///
/// Firestore'daki `config/announcement` belgesinden okunur, yani yeni bir
/// derleme yüklemeden duyuru yayınlanabilir. Bildirim göndermenin aksine
/// hiçbir Apple kurulumu (APNs anahtarı, izin ekranı) gerektirmiyor; buna
/// karşılık yalnızca oyunu AÇAN kişiye ulaşır. İkisi birbirinin yerine değil,
/// yanına: bildirim uzaktakini çağırır, bu kart geleni karşılar.
///
/// Belge alanları:
///   enabled     Bool    — kapalıysa hiçbir şey gösterilmez
///   id          String  — "kapat" takibi bunun üzerinden; metni değiştirip
///                         id'yi de değiştirirsen kapatanlar tekrar görür
///   title       String  — İngilizce taban
///   body        String
///   title_tr / body_tr / title_de / ... — dile göre karşılık (varsa)
///   minVersion  String  — "2.1". Uygulama bu sürümdeyse ya da üstündeyse
///                         duyuru GÖSTERİLMEZ. Güncelleme duyurusu için:
///                         güncellemeyi almış olan kişi çağrıyı görmemeli.
///   appStoreID  String  — düğme App Store'da bu kimliği açar; yoksa düğme yok
@MainActor
final class Announcement: ObservableObject {
    static let shared = Announcement()

    struct Item: Equatable {
        let id: String
        let title: String
        let body: String
        let appStoreID: String?
    }

    @Published private(set) var current: Item?

    private var didLoad = false
    private static func dismissKey(_ id: String) -> String { "lumo.announcement.seen.\(id)" }

    /// Uygulamanın kendi sürümü ("2.0")
    static var appVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0"
    }

    private init() {}

    /// Ana menü açılınca çağrılır. Oturumda bir kez okur — duyuru her açılışta
    /// yeniden indirilecek kadar önemli değil, ama uygulama yeniden başlarsa
    /// tazelenir.
    func refreshIfNeeded() async {
        guard !didLoad else { return }
        didLoad = true
        #if canImport(FirebaseCore)
        guard let raw = await FirebaseBridge.fetchAnnouncement() else { return }
        current = Self.resolve(raw, appVersion: Self.appVersion,
                               language: Locale.preferredLanguages.first ?? "en")
        #endif
    }

    /// Ham alanlardan gösterilecek duyuruyu çıkarır. Saf fonksiyon: sürüm
    /// karşılaştırması ve dil seçimi Firebase'e bağlanmadan da denenebilsin.
    static func resolve(_ data: [String: Any], appVersion: String,
                        language: String) -> Item? {
        guard data["enabled"] as? Bool ?? true else { return nil }
        guard let id = data["id"] as? String, !id.isEmpty else { return nil }
        guard !UserDefaults.standard.bool(forKey: dismissKey(id)) else { return nil }

        // Güncelleme çağrısı: hedeflenen sürüme ulaşmış olana gösterme
        if let minVersion = data["minVersion"] as? String, !minVersion.isEmpty,
           compare(appVersion, minVersion) >= 0 { return nil }

        // Dil kodu "tr-TR" gelir; ilk parçası yeter
        let lang = language.split(separator: "-").first.map(String.init) ?? "en"
        func field(_ name: String) -> String? {
            (data["\(name)_\(lang)"] as? String) ?? (data[name] as? String)
        }
        guard let title = field("title"), !title.isEmpty,
              let body = field("body"), !body.isEmpty else { return nil }

        let appStoreID = (data["appStoreID"] as? String).flatMap { $0.isEmpty ? nil : $0 }
        return Item(id: id, title: title, body: body, appStoreID: appStoreID)
    }

    /// "2.10" > "2.9" olmalı: parça parça SAYI karşılaştırması, metin değil.
    /// Düz string karşılaştırması "2.10" < "2.9" derdi.
    static func compare(_ a: String, _ b: String) -> Int {
        let lhs = a.split(separator: ".").map { Int($0) ?? 0 }
        let rhs = b.split(separator: ".").map { Int($0) ?? 0 }
        for i in 0..<max(lhs.count, rhs.count) {
            let l = i < lhs.count ? lhs[i] : 0
            let r = i < rhs.count ? rhs[i] : 0
            if l != r { return l < r ? -1 : 1 }
        }
        return 0
    }

    func dismiss() {
        guard let item = current else { return }
        UserDefaults.standard.set(true, forKey: Self.dismissKey(item.id))
        current = nil
    }

    /// App Store'daki sayfayı açar. `itms-apps` doğrudan App Store uygulamasına
    /// gider; https bağlantısı önce Safari'yi açıp oradan sıçrıyor.
    func openStore() {
        guard let id = current?.appStoreID,
              let url = URL(string: "itms-apps://itunes.apple.com/app/id\(id)") else { return }
        UIApplication.shared.open(url)
    }
}

File: android/app/src/main/java/com/caganhatapci/orbeon/services/Announcement.kt
package com.caganhatapci.orbeon.services

import android.content.Context
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.google.firebase.firestore.FirebaseFirestore
import java.util.Locale

/**
 * Ana menüde gösterilen duyuru kartı — "yeni sürüm çıktı", "bu hafta çift
 * yıldız" gibi. iOS'takiyle AYNI Firestore belgesini okuyor, yani bir kez
 * yazılan duyuru iki platformda birden görünüyor.
 *
 * Belge alanları:
 *   enabled     Bool    — kapalıysa hiçbir şey gösterilmez
 *   id          String  — "kapat" takibi bunun üzerinden
 *   title/body  String  — İngilizce taban
 *   title_tr / body_tr / …  — dile göre karşılık (varsa)
 *   minVersion  String  — uygulama bu sürümdeyse ya da üstündeyse GÖSTERİLMEZ
 *   playStoreId String  — düğme Play'de bu paketi açar; yoksa düğme yok
 */
class Announcement(private val context: Context) {

    data class Item(
        val id: String,
        val title: String,
        val body: String,
        val storeId: String?
    )

    var current by mutableStateOf<Item?>(null)
        private set

    private var didLoad = false
    private val p = context.getSharedPreferences("orbeon", Context.MODE_PRIVATE)

    /**
     * Menü açılınca çağrılır. Oturumda bir kez okuyor — duyuru her açılışta
     * yeniden indirilecek kadar önemli değil, ama uygulama yeniden başlarsa
     * tazeleniyor.
     */
    fun refreshIfNeeded(appVersion: String, push: PushManager?) {
        if (didLoad) return
        didLoad = true
        runCatching {
            FirebaseFirestore.getInstance().collection("config").document("announcement")
                .get()
                .addOnSuccessListener { snap ->
                    val data = snap.data ?: return@addOnSuccessListener
                    current = resolve(data, appVersion, Locale.getDefault().language)
                    // Aynı haber birkaç saat sonra bildirim olarak da düşsün:
                    // kartı görüp "sonra" diyene ikinci bir dokunuş.
                    push?.syncUpdateReminder(current?.title, current?.body)
                }
        }.onFailure { Log.w("Orbeon.Announcement", "Duyuru okunamadı: ${it.message}") }
    }

    /**
     * Ham alanlardan gösterilecek duyuruyu çıkarır. Saf fonksiyon: sürüm
     * karşılaştırması ve dil seçimi Firebase'e bağlanmadan da denenebilsin.
     */
    fun resolve(data: Map<String, Any?>, appVersion: String, language: String): Item? {
        if (data["enabled"] as? Boolean == false) return null
        val id = data["id"] as? String ?: return null
        if (id.isBlank()) return null
        if (p.getBoolean("announcement.seen.$id", false)) return null

        // Güncelleme çağrısı: hedeflenen sürüme ulaşmış olana gösterme
        val minVersion = data["minVersion"] as? String
        if (!minVersion.isNullOrBlank() && compare(appVersion, minVersion) >= 0) return null

        val lang = language.lowercase().split("-").first()
        fun field(name: String): String? =
            (data["${name}_$lang"] as? String) ?: (data[name] as? String)

        val title = field("title")?.takeIf { it.isNotBlank() } ?: return null
        val body = field("body")?.takeIf { it.isNotBlank() } ?: return null
        val storeId = (data["playStoreId"] as? String)?.takeIf { it.isNotBlank() }
        return Item(id, title, body, storeId)
    }

    /**
     * "2.10" > "2.9" olmalı: parça parça SAYI karşılaştırması, metin değil.
     * Düz string karşılaştırması "2.10" < "2.9" derdi.
     */
    fun compare(a: String, b: String): Int {
        val lhs = a.split(".").map { it.toIntOrNull() ?: 0 }
        val rhs = b.split(".").map { it.toIntOrNull() ?: 0 }
        for (i in 0 until maxOf(lhs.size, rhs.size)) {
            val l = lhs.getOrElse(i) { 0 }
            val r = rhs.getOrElse(i) { 0 }
            if (l != r) return if (l < r) -1 else 1
        }
        return 0
    }

    fun dismiss(push: PushManager?) {
        val item = current ?: return
        p.edit().putBoolean("announcement.seen.${item.id}", true).apply()
        current = null
        // Kartı kapatan haberi almış demektir; bildirimi de düşürmek aynı
        // şeyi ikinci kez söylemek olurdu
        push?.syncUpdateReminder(null, null)
    }
}
